### PR TITLE
Feature(PHP-117)- delete list

### PR DESCRIPTION
**Pull request by G4MAlexVannorsdall**
## Description
This step is apart of requirements Part 4 on the Google Doc. For this requirement I need to add the feature of being able to delete a shopping list. 

## How to run
Checkout as normal, using the terminal.

## Implementation
Had a discussion with you and looked through old code. Also created a short blue print of what I would need to do for this step.

## Thoughts/Considerations
Each time I push the code it will be a small micro step until the entire step is complete.

### CODE REVIEW

File: src/Application/Commands/DeleteShoppingList/DeleteShoppingListCommand.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    /**
+     * @param ShoppingListRepositoryInterface $repository
+     */
+    public function __construct(ShoppingListRepositoryInterface $repository)
+    {
+        $this->repository = $repository;
+    }
+
+    public function execute(DeleteShoppingListModel $model): void
```
Comment (lindyhopchris) [inline src/Application/Commands/DeleteShoppingList/DeleteShoppingListCommand.php:25]: Required: docblock

File: src/Application/Commands/DeleteShoppingList/DeleteShoppingListCommand.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        $list = $this->repository->findOrFail(
```
Comment (lindyhopchris) [inline src/Application/Commands/DeleteShoppingList/DeleteShoppingListCommand.php]: Comment: this morning we talked about whether it was necessary to load the entity from storage to then delete it.

Comment (G4MAlexVannorsdall) [reply src/Application/Commands/DeleteShoppingList/DeleteShoppingListCommand.php]: Okay, I updated this file using the fileHandler. Once the fileHandler has the delete/unlink method in it then I could use it to delete the file with the given slug. For right now the updated code in this file is seeing if the file with the given slug exists.

File: src/Application/Commands/DeleteShoppingList/DeleteShoppingListCommand.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /**
+     * @var JsonFileHandler
+     */
+    private JsonFileHandler $fileHandler;
+
+    /**
+     * @param JsonFileHandler $fileHandler
+     */
+    public function __construct(JsonFileHandler $fileHandler)
```
Comment (lindyhopchris) [inline src/Application/Commands/DeleteShoppingList/DeleteShoppingListCommand.php]: Required: this change to the constructor dependency isn't right I'm afraid. The purpose of the repository interface is that it decouples the commands from a _specific_ storage implementation. By injecting the `JsonFileHandler` here, your command will only ever work with the JSON storage. If we switched to using a database storage, then this command wouldn't work.

This morning we talked about implementing a `delete` method on the repository interface. You need to stick to that approach. Internally, the JSON storage (that implements the interface) would then delete the file via the file handler.

File: src/Application/Commands/DeleteShoppingList/DeleteShoppingListCommand.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        $this->repository = $repository;
+    }
+
+    /**
+     * @param DeleteShoppingListModel $model
+     */
+    public function execute(DeleteShoppingListModel $model): void
+    {
+        // Get the list
+        $list = $this->repository->findOrFail(
```
Comment (lindyhopchris) [inline src/Application/Commands/DeleteShoppingList/DeleteShoppingListCommand.php]: Tip: we talked about not needing to load the shopping list entity, just to then delete it. We might need to discuss that again here, as you are currently loading then deleting the entity.

File: src/Domain/ValueObjects/Slug.php [hunk 5]
```diff
@@ -6,7 +6,7 @@
 use JsonSerializable;
 use UnexpectedValueException;
 
-class Slug implements JsonSerializable
+class Slug extends \Lindyhopchris\ShoppingList\Domain\ShoppingList implements JsonSerializable
```
Comment (lindyhopchris) [inline src/Domain/ValueObjects/Slug.php]: Question: wondering why this change has been made?

File: src/Persistance/Json/JsonShoppingListRepository.php [hunk 6]
```diff
@@ -90,4 +90,14 @@ private function storeAs(ShoppingList|string $slug): string
 
         return $slug . '.json';
     }
+
+    public function unlink(ShoppingList $slug): void
+    {
+        $filename = $this->storeAs($slug);
+
+        if ($this->files->exists($filename)) {
+            $this->unlink($slug);
```
Comment (lindyhopchris) [inline src/Persistance/Json/JsonShoppingListRepository.php]: Question: this will cause an infinite loop. Can you identify why?

File: src/Persistance/ShoppingListRepositoryInterface.php [hunk 7]
```diff
@@ -39,4 +39,12 @@ public function findOrFail(string $slug): ShoppingList;
...
      */
     public function store(ShoppingList $list): void;
+
+    /**
+     * Retrieve a shopping list by its slug and delete it.
+     *
+     * @param ShoppingList $slug
+     * @return void
+     */
+    public function unlink(Shoppinglist $slug): void;
```
Comment (lindyhopchris) [inline src/Persistance/ShoppingListRepositoryInterface.php]: Required: `unlink` has a specific meaning to it in PHP, which is to remove a file. The repository interface should use language that is not specific to the underlying implementation. Words like `delete` or `remove` would be more descriptive of what the calling code is expecting to happen when this method is called, regardless of what the underlying technical implementation has to do to physically remove the entity.

Required: the type-hint isn't correct here. Your parameter is called `$slug` but you've type-hinted it as `ShoppingList`. As mentioned, we did discuss this on Monday - we don't need the full shopping list to delete it, we probably only need the slug (aka unique identifier for the list).

File: src/Persistance/ShoppingListRepositoryInterface.php [hunk 8]
```diff
@@ -39,4 +39,12 @@ public function findOrFail(string $slug): ShoppingList;
      * @return void
      */
     public function store(ShoppingList $list): void;
+
+    /**
+     * Retrieve a shopping list by its slug and delete it.
```
Comment (lindyhopchris) [inline src/Persistance/ShoppingListRepositoryInterface.php]: Tip: the description of this method isn't accurate for what this method should do - it should delete the shopping list. I.e. the "retrieve" isn't necessary.

File: src/Application/Commands/DeleteShoppingList/DeleteShoppingListCommand.php (hunk 4 above)
Comment (G4MAlexVannorsdall) [reply src/Application/Commands/DeleteShoppingList/DeleteShoppingListCommand.php]: What got me is I feel like I am supposed to use the Delete model yet here that load the entity? I am confused on that I will admit.

File: src/Persistance/Json/JsonFileHandler.php [hunk 9]
```diff
@@ -109,6 +109,25 @@ public function write(string $filename, mixed $jsonable): void
...
+     *
+     * @param string $filename
+     * @return void
+     */
+    public function unlink(string $filename): void
+    {
+        // Given the path to the file using the slug
+        $slug = $this->pathTo($filename);
+        // If the list of that filename exists then
+        if ($this->exists($filename) && $filename === $slug) {
```
Comment (lindyhopchris) [inline src/Persistance/Json/JsonFileHandler.php]: Required: this won't work, but is extremely close. The comment you've put above the `if` statement does accurately describe what the if should check, but what you've written in code doesn't match it.

File: src/Persistance/Json/JsonFileHandler.php [hunk 10]
```diff
@@ -109,6 +109,25 @@ public function write(string $filename, mixed $jsonable): void
...
+    /**
+     * Remove the file from storage.
+     *
+     * @param string $filename
+     * @return void
+     */
+    public function unlink(string $filename): void
+    {
+        // Given the path to the file using the slug
+        $slug = $this->pathTo($filename);
```
Comment (lindyhopchris) [inline src/Persistance/Json/JsonFileHandler.php]: Suggestion: `$slug` is an inaccurate name for this variable, because the `pathTo()` method doesn't return the slug - it returns the full path to the file. Take a look at what the variable is called in the other methods to get an idea of what it could be called.

File: src/Persistance/Json/JsonShoppingListRepository.php [hunk 11]
```diff
@@ -90,4 +91,18 @@ private function storeAs(ShoppingList|string $slug): string
...
+
+    /**
+     * @param string $slug
+     * @return void
+     */
+    public function delete(string $slug): void
+    {
+        $filename = $this->storeAs($slug);
+
+        if ($this->files->exists($filename)) {
```
Comment (lindyhopchris) [inline src/Persistance/Json/JsonShoppingListRepository.php]: Suggestion: you're checking if the file exists here, but the file handler also checks whether the file exists. So that means this check is being done twice - it would be more efficient to check whether it exists just once. There's no right answer as to whether the repo or the file handler is the best place to check if the file exists - different developers would go for different approaches. Personally I'd check in the file handler i.e. the repo can just call `$this->files->unlink()` without the `if` statement.

**Review by lindyhopchris — CHANGES_REQUESTED** [review]
Looks good! Just a few little things to tidy up, then we need to think about unit tests.

File: src/Persistance/Json/JsonFileHandler.php (hunk 9 above)
Comment (G4MAlexVannorsdall) [reply src/Persistance/Json/JsonFileHandler.php]: I updated the code!

File: src/Persistance/Json/JsonFileHandler.php (hunk 10 above)
Comment (G4MAlexVannorsdall) [reply src/Persistance/Json/JsonFileHandler.php]: I updated the variable name to $path.

File: src/Persistance/Json/JsonShoppingListRepository.php (hunk 11 above)
Comment (G4MAlexVannorsdall) [reply src/Persistance/Json/JsonShoppingListRepository.php]: Ah, that does make sense. I chose the approach you have stated above and just simply did $this->files-unlink($filename) in the repo. It makes sense that the fileHandler would check if the file exists since it handles everything that deals with the files.

File: src/Persistance/Json/JsonFileHandler.php [hunk 12]
```diff
@@ -109,6 +109,25 @@ public function write(string $filename, mixed $jsonable): void
...
+     * @return void
+     */
+    public function unlink(string $filename): void
+    {
+        // Given the path to the file
+        $path = $this->pathTo($filename);
+        // If the list of that filename exists then
+        if ($this->exists($filename)) {
+            // Delete the list
+            unlink($filename);
```
Comment (lindyhopchris) [inline src/Persistance/Json/JsonFileHandler.php]: Required: when unlinking, you need to use the full path to the filename. As an explanation of what's going on here, `$filename` is just the filename - e.g. `my-groceries.json`. When you call `unlink()` PHP needs to know exactly where on the filesystem the file is to delete. So rather than passing it `my-groceries.json` via the `$filename` variable, you need to use the full path which would be something like `/Users/alex.vannorsdall/code/php-shopping-list/storage/my-groceries.json`. That's what the `pathTo()` method does in this file handler class - it takes the filename and adds the directory to return a string that is the exact location on the filesystem of that file.

Comment (G4MAlexVannorsdall) [reply src/Persistance/Json/JsonFileHandler.php]: Alright, the does make sense! Thank you for the explanation!

File: tests/Unit/Persistence/Json/JsonShoppingListRepositoryTest.php [hunk 13]
```diff
@@ -155,4 +155,16 @@ public function testStore(): void
 
         $this->repository->store($list);
     }
+
+    public function testDelete(): void
+    {
+        // Arrange: Given a shopping list
+        // Act: We expect the list to be called on once, using the delete method with the list
+        $this->files
```
Comment (lindyhopchris) [inline tests/Unit/Persistence/Json/JsonShoppingListRepositoryTest.php:163]: Required: this test doesn't accurately test the scenario. When a file exists, you're expecting the `exists` method to be called first and return `true`. Then you're expecting `unlink` to be called. 

So the correct code here for the test is:

```php
$this->files
    ->expects($this->once())
    ->method('exists')
    ->with('supplies.json')
    ->willReturn(true);

$this->files
    ->expects($this->once())
    ->method('unlink')
    ->with('supplies.json');

$this->repository->delete('supplies');
```

This is equivalent to:

- Given a shopping list called "supplies" that exists;
- When the list is deleted via the repository;
- Then the file handler unlinks (deletes) the file.

This only tests what happens when deleting a list that exists - we should also test deleting a list that does not exist.

- Given that a shopping list called "supplies" does not exist
- When the list is deleted via the repository
- Then an exception is thrown telling us that the list does not exist.

Have a go at adding that as a new test on your `JsonShoppingListRepositoryTest` class. You'll see that it fails, because in your last commit you deleted code in the JSON repository that was correct.

```php
$this->files
    ->expects($this->once())
    ->method('exists')
    ->with('supplies.json')
    ->willReturn(false);

$this->files
    ->expects($this->never())
    ->method('unlink');

$this->expectException(ShoppingListNotFoundException::class);

$this->repository->delete('supplies');
```

**Review by lindyhopchris — CHANGES_REQUESTED** [review]

File: tests/Unit/Persistence/Json/JsonShoppingListRepositoryTest.php (hunk 13 above)
Comment (G4MAlexVannorsdall) [reply tests/Unit/Persistence/Json/JsonShoppingListRepositoryTest.php:163]: I went through this feedback (thank you for this) and it does make sense. However both tests still don't pass. I got confused by what you meant was correct with the code because of the screenshot I sent you last week and I assumed that was right, but I went back in my commits and changed it again. 

My question is since the handler checks if they list exists do I still have to test to see if the list exists in the first test for the JsonShoppingListRepo?

Comment (G4MAlexVannorsdall) [reply tests/Unit/Persistence/Json/JsonShoppingListRepositoryTest.php:163]: Thank you for leaving back this feedback. I looked through it yesterday when it was in the middle of peak hotness and couldn't focus properly. I managed to fix the code so it correlated with the tests and both pass now.

**Review by lindyhopchris — APPROVED** [review]
Looks good!
